pkg/tnf/pacemaker/healthcheck.go

```diff
@@ -0,0 +1,718 @@
+package pacemaker
+
+import (
+	"context"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/health"
+	"github.com/openshift/cluster-etcd-operator/pkg/tnf/pkg/exec"
+)
+
+// Constants for time windows and status strings
+const (
+	// Time windows for detecting recent events
+	failedActionTimeWindow = 5 * time.Minute
+	fencingEventTimeWindow = 24 * time.Hour
+
+	// Resync interval for health check controller
+	healthCheckResyncInterval = 30 * time.Second
+
+	// Event deduplication windows - avoid recording the same event within these times
+	eventDeduplicationWindowFencing = 1 * time.Hour   // Fencing events create reminders every hour for 24 hours
+	eventDeduplicationWindowDefault = 5 * time.Minute // Resource actions and general warnings
+
+	// Timeout for executing pcs status xml command
+	execTimeout = 10 * time.Second
+
+	// Expected number of nodes in ExternalEtcd cluster
+	expectedNodeCount = 2
+
+	// Status strings
+	statusHealthy = "Healthy"
+	statusWarning = "Warning"
+	statusError   = "Error"
+	statusUnknown = "Unknown"
+
+	// Pacemaker state
+	pacemakerStateRunning = "running"
+
+	// Resource names
+	resourceKubelet = "kubelet"
+	resourceEtcd    = "etcd"
+
+	// Resource agent names
+	resourceAgentKubelet = "systemd:kubelet"
+	resourceAgentEtcd    = "ocf:heartbeat:podman-etcd"
+
+	// Node status indicators
+	nodeStatusStarted = "Started"
+
+	// Operation status indicators
+	operationRCSuccess     = "0"
+	operationRCTextSuccess = "ok"
+
+	// Quorum and active status
+	booleanValueTrue  = "true"
+	booleanValueFalse = "false"
+
+	// Degraded condition reason
+	reasonPacemakerUnhealthy = "PacemakerUnhealthy"
+
+	// Maximum XML size to prevent XML bombs (10MB)
+	maxXMLSize = 10 * 1024 * 1024
+
+	// Time formats for parsing Pacemaker timestamps
+	// Go's reference time: Mon Jan 2 15:04:05 MST 2006 (mnemonic: 1 2 3 4 5 6 7)
+	pacemakerTimeFormat      = "Mon Jan 2 15:04:05 2006"     // Format for operation history timestamps
+	pacemakerFenceTimeFormat = "2006-01-02 15:04:05.000000Z" // Format for fence event timestamps
+
+	// Pacemaker commands
+	pcsStatusXMLCommand = "sudo -n pcs status xml"
```

Comment:
Huh wait... did you already try running this in cluster?
IIRC one of the reasons for using jobs for pcs commands was about permissions... 🤔

Reply:
So it doesn't work in an interesting way.
```
[ec2-user@aws-jpoulin-dev ~]$ oc logs pod/etcd-operator-554d7f6899-5wrp2 | grep health.go
E1015 15:16:40.477459       1 health.go:120] health check for member (master-0) failed: err(context deadline exceeded)
E1015 15:16:45.703453       1 health.go:120] health check for member (master-0) failed: err(context deadline exceeded)
E1015 15:19:04.918498       1 health.go:120] health check for member (master-1) failed: err(context deadline exceeded)
```

Why does it stop after 3 tries I wonder :P
I'll try to update via chron with privileges and see if I can get something working